crates/wasmparser/src/validator/core.rs

```diff
@@ -956,20 +961,23 @@ impl Module {
             | HeapType::Eq
             | HeapType::Struct
             | HeapType::Array
-            | HeapType::I31 => Ok(()),
-            HeapType::Concrete(type_index) => {
-                match type_index {
-                    UnpackedIndex::Module(idx) => {
-                        let _ = self.type_id_at(idx, offset)?;
-                        Ok(())
-                    }
-                    UnpackedIndex::RecGroup(_) | UnpackedIndex::Id(_) => {
-                        // If the type index has already been canonicalized,
-                        // then we already checked that it was in bounds and
-                        // valid at that time.
-                        Ok(())
-                    }
-                }
+            | HeapType::I31 => return Ok(()),
+            HeapType::Concrete(type_index) => type_index,
+        };
+        match type_index {
+            UnpackedIndex::Module(idx) => {
+                let id = self.type_id_at(*idx, offset)?;
+                *type_index = UnpackedIndex::Id(id);
+                Ok(())
+            }
+            UnpackedIndex::RecGroup(_) => {
+                todo!()
+            }
+            UnpackedIndex::Id(_) => {
+                // If the type index has already been canonicalized,
+                // then we already checked that it was in bounds and
+                // valid at that time.
+                Ok(())
```

Comment:
These should both be unreachable after this PR, right? Since the idea is that we check once and that does the canonicalization, and then we never check again?

Reply:
Ah yes good point!